!!! FEATURE:  Global variables for the Fusion runtime `FusionGlobals`

mhsdesign opened the pull request:
<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->


```php
$runtime = $this->runtimeFactory->createFromConfiguration(
    $this->createFusionConfigurationFromSite($this->findSiteBySiteNode($currentSiteNode)),
    FusionGlobals::fromArray(
        [
            'request' => $actionRequest',
            'any' => "thing",
            'Foo' => [
               'SomeHelper' => $someHelper
            ]
        ]
);
// optional: $runtime->setControllerContext($controllerContext);
```

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
!!! FEATURE: Fusion runtime: Custom default context variables

mhsdesign committed:
WIP: Runtime dont require `request` and calculate `controllerContext` from it

Fix some tests rudimentary and adjust RuntimeFactory

mficzel commented on Neos.Neos/Classes/Fusion/FusionVariableAccessor.php:
```diff
@@ -0,0 +1,53 @@
+ * But in the context of the Neos Fusion rendering, we assume certain variables like `request` and `node`.
+ * This accessor provides a type safe access to the conventional variables.
+ */
+final class FusionVariableAccessor
```
As discussed this should be a seperate pr

mficzel reviewed (commented):
Pretty sure the `\Neos\Fusion\View` has to be adjusted aswell.
Other than that i get the point and agree this.

mhsdesign commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -227,6 +270,15 @@ public function getCurrentContext()
         return $this->currentContext;
     }
 
+    /**
+     * @param string $name the key of the variable like `request` or `node`.
+     * @return mixed null if not set
+     */
+    public function getContextVariable(string $name): mixed
+    {
+        return $this->currentContext[$name] ?? $this->defaultContextVariables->value[$name] ?? null;
+    }
```
Alternatively (or rather additionally) suggested: 

```php
public function getContextObjectOfType(string $name, string $classStringName): object
{
    $value = $this->getContextVariable($name, $classStringName);
    if (!$value instanceof ActionRequest) {
        throw new \RuntimeException(sprintf('Expected Fusion variable "request" to be of type ActionRequest, got value of type "%s".', $classStringName, get_debug_type($value)), 1693557504154);
    }
    return $value;

    return $this->currentContext[$name] ?? $this->defaultContextVariables->value[$name] ?? null;
}
```

mhsdesign replied:
Will be discussed here https://github.com/neos/neos-development-collection/issues/4475

mhsdesign committed:
TASK: Adjust tests to avoid `RuntimeFactory::create` and remove $controllerContext

mhsdesign commented on Neos.Fusion/Classes/Core/RuntimeFactory.php:
```diff
@@ -39,24 +45,38 @@ public function create(array $fusionConfiguration, ControllerContext $controller
-    public function createFromConfiguration(FusionConfiguration $fusionConfiguration, ControllerContext $controllerContext): Runtime
-    {
-        return new Runtime($fusionConfiguration, $controllerContext);
+    public function createFromConfiguration(
+        FusionConfiguration $fusionConfiguration,
+        FusionDefaultContextVariables $additionalDefaultContextVariables
+    ): Runtime {
```
todo: discuss signature with @bwaidelich to be forward compatible to new action uri building stuff https://github.com/neos/flow-development-collection/pull/3002#issuecomment-1503762615

mhsdesign replied:
we could in the future possibly do this:

```php
FusionDefaultContextVariables::fromArray(["actionUriBuilder" => $builder, "request" => $request])
```

pass the `actionUriBuilder` also as runtime default context variable - that would allow us to omit the `request` for instantiation of the runtime, and the uribuilding inside of fusion will still work. The only downside would be that the variable is accessible via eel, which might be odd but doing no harm: `${actionUriBuilder}`

mhsdesign committed:
TASK: Replace `RuntimeFactory::create`

mhsdesign committed:
Revert introduction of `FusionVariableAccessor`

mhsdesign commented on Neos.Fusion/Classes/Core/FusionDefaultContextVariables.php:
```diff
@@ -0,0 +1,37 @@
+ * Variables from configuration that should be set in the context by default.
+ * For example Eel helper definitions
+ */
+final class FusionDefaultContextVariables
```
Im not totally happy with the naming here.

It should ideally reflect that these are the variables / options / metadata / properties ... that are valid for this specific runtime instance. They are not cached and "should ideally not be overridden - at least conceptional". Their purpose is to hold eel helpers or things like `request` and the upcoming editPreviewMode.

They are different to the variables like "node", as they can be set at runtime.

But they are similar to "node" and other variables, as they can be accessed like regulary in eel or via `$runtime->getContextVariable`

mhsdesign replied:
@grebaldi had the idea to name it `FusionGlobals` and i really like it :D

mhsdesign commented on Neos.Fusion/Classes/Core/RuntimeFactory.php:
```diff
@@ -39,24 +45,38 @@ public function create(array $fusionConfiguration, ControllerContext $controller
+        FusionConfiguration $fusionConfiguration,
+        FusionDefaultContextVariables $additionalDefaultContextVariables
+    ): Runtime {
+        $defaultContextVariables = FusionDefaultContextVariables::fromArray(
+            EelUtility::getDefaultContextVariables(
+                $this->defaultContextConfiguration ?? []
+            )
+        );
+        return new Runtime($fusionConfiguration, $defaultContextVariables->merge($additionalDefaultContextVariables));
```
im currently enriching the `$additionalDefaultContextVariables` with the eel helpers from the config. instead of using two kinds of different dto classes (one only for the factory) - i think this makes the most sense?

mhsdesign replied:
yes after the rename to `FusionGlobals` it makes totally sense!

mficzel commented on Neos.Fusion/Classes/View/FusionView.php:
```diff
@@ -179,11 +187,9 @@ protected function loadFusion()
+     * Parse all the fusion files that are in the current fusionPathPatterns
      */
-    protected function getMergedFusionObjectTree(): array
+    protected function getMergedFusionObjectTree(): FusionConfiguration
```
Is that signature change really needed here? This is used by monocle and would require a new release for Neos 9.

mhsdesign replied:
Good catch. Was just cleaning up. The alternative from using the DTO through all layers is 

Parser -> DTO -> to array -> to DTO -> Runtime

mhsdesign replied:
It seems the method is fine as is, but changing the property type of $parsedFusion breaks?

https://github.com/sitegeist/Sitegeist.Monocle/blob/5b33c359f88251d9563de1a068d78c32daf87bfe/Classes/Fusion/FusionView.php#L49

mficzel replied:
Monocle calls this method aswell so the signature change there is also relevant.

mhsdesign replied:
Sorry I can’t find it ^^

mficzel replied:
You are right. Monocle calls its own method getMergedFusionObjectTreeForSitePackage but this had to change signature since parsedFusion changed the type.

Am not strictly against changing this but we already declared monocle Neos 9 compliant some time ago since we expected no changes in this regard as we adjusted the fusion parser stuff in 8.3 for that. 

It would be really great to be able to support Neos 8 and 9 from a single package version since i expect Neos 8 to be still relevant for a while.

mhsdesign replied:
Hmm i value your point and will not destroy your plan (be the reason for the extra hassle of a dedicated 90 version).
How about adding a tiny little bit of conditional code like:

```php
public function loadFusion() {
    if (FLOW_VERSION === '90') {
        // or a try catch, or actually asking the package manager
        // or inspection: $isTypedProperty = (new ReflectionProperty($this, 'parsedFusion'))->hasType();
        $this->parsedFusion = $someService->getMergedFusionObjectTreeForSitePackage();
    } else {
        $this->parsedFusion = $someService->getMergedFusionObjectTreeForSitePackage()->toArray();
    }
}
```

i do think its important to add some type safety to help people extending the view point in the right direction and prevent them from foot-guns.

mficzel replied:
Hmmm will try that here and use a custom FusionService:

```php
public function loadFusion() {
        $r = new \ReflectionClass(static::class);
        $parsedFusionType = $r->getProperty('parsedFusion')->getType();
        if ($parsedFusionType->getName() === FusionConfiguration::class) {
            $this->parsedFusion = $this->fusionService->getMergedFusionObjectTreeForSitePackage($this->getOption('packageKey'));
        } else {
            $this->parsedFusion = $this->fusionService->getMergedFusionObjectTreeForSitePackage($this->getOption('packageKey'))->toArray();
        }
}
```
Will tell you the result shortly

mficzel replied:
Seems work reasonably well here https://github.com/sitegeist/Sitegeist.Monocle/pull/221 ... will try in 8.3 but that would be fine for me

mficzel replied:
Ok i could adjust monocle to run on 8.3 and 9.0 so i consider this solved.

mficzel reviewed (commented):
-

mficzel commented on Neos.Fusion/Classes/View/FusionView.php:
```diff
@@ -54,10 +56,8 @@ class FusionView extends AbstractView
-     * @var array
      */
-    protected $parsedFusion;
+    protected FusionConfiguration $parsedFusion;
```
Changing this from array to FusionConfiguration is also breaky.

mhsdesign replied:
Yes, but we evaluated that to be okayish - see https://github.com/neos/neos-development-collection/pull/4425#discussion_r1319460251 for a workaround ;)

mficzel replied:
agreed that was solved

mficzel reviewed (changes requested):
Generally fine and i get the point:

1. \Neos\Neos\View\FusionView::getFusionRuntime should be adjusted this still uses fusionService->createRuntime that is marked for deletion in 9
2. Can we reduce the breakiness because the signature changes in the FusionView + FusionService  (we can deal with it but i would like to know wether it is really necessary
3. Regarding naming. How about `DefaultFusionContextVariablesThatMagicallyShineThroughCacheBoundaries` ... only half joking. If there is no good name it sometimes is best to use an descriptive name.

mhsdesign commented:
We discussed that we want to let the default context variables win over the fusion vars. So no overriding of `request` anymore.
Also in the "runtime system variables" we should avoid the wording "context"

mficzel commented:
> We discussed that we want to let the default context variables win over the fusion vars. So no overriding of request anymore.

is that tested with fusion runtime forms?

mhsdesign commented:
> is that tested with fusion runtime forms?

good catch hmmm // cc @bwaidelich

mficzel commented:
Here you see: https://github.com/neos/fusion-form/blob/master/Classes/Runtime/FusionObjects/RuntimeFormImplementation.php#L103

mficzel commented:
I would suggest to still adjust `\Neos\Fusion\View` to use the defaultContext and otherwise keep this pr as it is.

I am against preventing to override default contexts. The consequences of changing this are not well understood at this time and it is not necessary for this change. I personally know cases where `request` was adjusted (FusionForm) and where EelHelpers were replaced for specific parts (hacky but possible).

mficzel commented:
The new name fusionGlobals @grebaldi suggested is nice. How about defining that those values cannot simply be overwritten in fusion via @context but add an internal method to the runtime.

i will also check wether we can use a separate „formRequest“ context variable in fusion form. That would be breaky for all cases where the request is used to configure schemas dynamically but otherwise fine.

mhsdesign committed:
TASK: Rename FusionDefaultContextVariables to FusionGlobals and disallow override of variables

mhsdesign commented:
> How about defining that those values cannot simply be overwritten in fusion via @context but add an internal method to the runtime.

good idea. `pushContext` and `pushContextArray` will still allow it, but we document this to be "highly discouraged"

> Overriding Fusion globals is disallowed via \@context and highly discouraged via {@see Runtime::pushContext()}.

mficzel reviewed (changes requested):
I still would like to see the adjustments to `\Neos\Neos\View\FusionView` and `\Neos\Neos\View\FusionExceptionView` in this pr. That way the change in the editPreview mode pr will be about adding another global and not about adjusting the Views.

mhsdesign commented:
Hmm i was thinking, that those neos.neos views will be adjusted by your pr - as i dont have the UiMode dto class yet:

Neos.Neos FusionView
```
protected $supportedOptions = [
    'uiMode' => [null, 'The fusion global variable: ui mode.', UiMode::class]
];
```

I added the view option of the general concept of globals to the Neos.Fusion view - but imo this concept is to low level for the neos view

mhsdesign committed:
FEATURE: FusionView option "fusionGlobals"

mficzel commented:
Adding the dto and ui mode will be the next pr. This one should add using the new globalContext  with just „request“. That way the views are refactored from the deprecated `RuntimeFactory::create` to the new `RuntimeFactory::createFromConfiguration` that is introduced here. This is also why i think this should go here.

That way it is crystal clear where the uiMode will go in the next pr. This one will have enough changes on its own.

mficzel commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -568,6 +614,7 @@ protected function prepareContextForFusionObject(AbstractFusionObject $fusionObj
 
         if (isset($newContextArray)) {
+            $this->fusionGlobals->requireContextToNotInterfereWithGlobals($newContextArray);
             $this->pushContextArray($newContextArray);
```
Maybe change signature of pushContextArray to
```
function pushContextArray(array $newContextArray, bool $preventOverwritingGlobals = false): void
```
that way it would be backwards compatible and still confine the checks to the pushContextArray method.

mhsdesign replied:
hmm sorry i think this flag is to complex. Instead i could imagine a separate `pushContextAndOverrideGlobal` method - or keep everything as is on good trust basis ^^

mficzel replied:
`pushContextAndOverrideGlobals` would be fine for me aswell but the logic imho belongs into the runtime. Fusion.Form would have to get a flow version detection for that but this is ok.

mhsdesign replied:
Fusion.Form could imo as well use `method_exists($runtime, 'pushContextAndOverrideGlobals')` and based upon that call either that or fallback to `pushContext`.

And then yes i can move the `requireContextToNotInterfereWithGlobals` logic into `pushContext` ^^.

Additionally i would deprecate `pushContextAndOverrideGlobals` (without a removal) as its highly discouraged.

mficzel replied:
Add a note to the deprecation that this is currently in use by fusion form. Maybe annotation as internal would make sense. Not much opinion there from my side.

mhsdesign replied:
Implemented :D

mficzel commented on Neos.Fusion/Classes/Core/FusionGlobals.php:
```diff
@@ -0,0 +1,68 @@
+        );
+    }
+
+    public function requireContextToNotInterfereWithGlobals(array $context): void
```
Looks a bit strange to me in here. Would expect this logic in pushContext

mhsdesign committed:
TASK: Avoid deprecated FusionService::createRuntime

mhsdesign commented:
> Adding the dto and ui mode will be the next pr. This one should add using the new globalContext with just „request“. That way the views are refactored from the deprecated `RuntimeFactory::create` to the new `RuntimeFactory::createFromConfiguration` that is introduced here. This is also why i think this should go here.
> 
> That way it is crystal clear where the uiMode will go in the next pr. This one will have enough changes on its own.

yes sorry i misunderstood ^^ i adjusted the code.

mhsdesign committed:
TASK: Fusion runtime `pushContext` cannot override globals.

A legacy layer for Fusion Forms has been implemented `pushContextAndOverrideGlobals`

mhsdesign commented on Neos.Fusion/Classes/Core/FusionGlobals.php:
```diff
@@ -0,0 +1,58 @@
+    public function merge(FusionGlobals $other): self
+    {
+        return new self(
+            Arrays::arrayMergeRecursiveOverrule($this->value, $other->value)
```
we only need a recursive merge, if we want to allow that one can set a helper inside the Neos array from outside and it will be merged ... i think its okayish and does no harm for now :D

mhsdesign commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -94,21 +98,19 @@ class Runtime
+     * Fusion global variables like EEL helper definitions {@see FusionGlobals}
      */
-    protected $defaultContextVariables;
+    public readonly FusionGlobals $fusionGlobals;
```
fyi public readonly!

mhsdesign commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -179,25 +222,47 @@ public function addCacheTag($key, $value)
      */
     public function pushContextArray(array $contextArray)
     {
+        foreach ($this->fusionGlobals->value as $disallowedFusionContextKey => $_) {
+            if (array_key_exists($disallowedFusionContextKey, $contextArray)) {
+                throw new RuntimeException(sprintf('Overriding Fusion global variable "%s" via @context is not allowed.', $disallowedFusionContextKey), 1694247627130);
+            }
+        }
```
@mficzel or is it faster to first iterate over the `$contextArray` -> would be if it has less entries than `$this->fusionGlobals->value`

mficzel replied:
I think it would be way faster to check the keys that were used in `$fusionConfiguration['__meta']['context']` here: 
https://github.com/neos/neos-development-collection/blob/9.0/Neos.Fusion/Classes/Core/Runtime.php#L563C14-L568. Similar to what you did in `pushContext`. That way we check only the newly added context keys and not everything. That would still prevent any global overriding from fusion and `pushContext`but leave  `pushContextArray` open for "creative" use via php. 

`pushContextArray` is described as beeing internal anyways and should get that as official annotation. So i would suggest to only add the check `pushContext` and the handling of `['__meta']['context']` and leave `pushContextArray` as it is as an internal footgun. 

I can can adjust Fusion.Form to set the `request` via `pushContextArray` so we need no special rules there. As this is in the Fusion namespace i consider using internal api fine there.

mhsdesign replied:
I like this approach 👍

mhsdesign replied:
Thanks for the rich discussions ^^, its now implemented as you described.

mhsdesign committed:
TASK: Fusion runtime `pushContextArray` can override globals.

Reverts partially 551b9531b9a30bdcff082ec97ee00cfbc58f61c0

mhsdesign commented:
Fyi @mficzel i did a little follow clean up, as i couldn't brace myself after seeing the FusionView.
Here is a cleanup which should make thinks more fun for the developer: https://github.com/neos/neos-development-collection/pull/4495

mficzel commented on Neos.Fusion/Classes/Core/FusionGlobals.php:
```diff
@@ -0,0 +1,58 @@
+     * You can access the current request like via this getter:
+     * `$runtime->fusionGlobals->getGlobal('request')`
+     */
+    public function getGlobal(string $name): mixed
```
Is there a reason this is not simply `get`

mhsdesign replied:
no i rolled i dice :D - but now you say it and its 2 vs 1 vote so yes (my brain liked both 😂 )

mficzel commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -191,14 +237,18 @@ public function pushContextArray(array $contextArray)
      */
     public function pushContext($key, $context)
     {
+        if (array_key_exists($key, $this->fusionGlobals->value)) {
```
one could also write this as
```
if ($this->fusionGlobals->hasKey($key))
```

mhsdesign replied:
how did i miss that ??? -> thank you good idea :D

mficzel commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -563,6 +615,9 @@ protected function prepareContextForFusionObject(AbstractFusionObject $fusionObj
         if (isset($fusionConfiguration['__meta']['context'])) {
             $newContextArray ??= $this->currentContext;
             foreach ($fusionConfiguration['__meta']['context'] as $contextKey => $contextValue) {
+                if (array_key_exists($contextKey, $this->fusionGlobals->value)) {
```
same here possibly:
```
if ($this->fusionGlobals->hasKey($contextKey)
```

mficzel reviewed (changes requested):
I like the change but get errors while testing locally in the backend end frontend. Does this work on your machine?

Other than that i would only suggest to adjust some nitpics.

mficzel reviewed (approved):
Got it to work after:

1. Merging the 9.0 branch of Neos (to be in line with change in the ui)-
2. Checkout PR https://github.com/neos/fusion-form/pull/89 of Neos.Fusion.Form !!! This is necessary since the startpage of Neos.Demo shows a Form !!!
3. Checkout PR https://github.com/sitegeist/Sitegeist.Monocle/pull/221 of Sitegeist.Monocle (if you use Monocle)

Still would love to see the little nitpicks addressed.

mhsdesign committed:
TASK: Add `FusionGlobals` `get` and `has`

mhsdesign commented:
Thank you for making sure it works on your setup and also thank you for your quick reviews!!

Even if we have formally here only 1 green review - as we discussed this topic within a bigger group multiple times and agreed on this solution, i will merge this after the tests are green.

mficzel commented:
Jiha … thanks for the effort …. will start working on the next iteration of the edit preview modes shortly. Might take a few days to find the needed hours
